[v7r3] register args (optional)

TaykYoku opened the pull request:
Example:
```
"""My script

Usage:
  dirac-myscript [options] ... Status <DN|ID> Element [Element]

Arguments:
  Status:          element status ['done', 'fail']
  DN:               user DN
  ID:                 user ID
  Element:       list of elements [default: 'no elements'] (optional)
"""
from DIRAC.Core.Base import Script
from DIRAC.Core.Utilities.DIRACScript import DIRACScript

@DIRACScript()
def main():
  Script.parseCommandLine()
  args = Script.getPositionalArguments()
  
  # Check if all mandatory arguments is present and the value of the first argument corresponds to the specified value
  if len(args) < 2 or (args[0] not in ['done', 'fail']):
     Script.showHelp()

  status = args[0]

  # detection of a controversial argument
  if args[1].startswith('/'):
    dn = args[1]
    id = None
  else:
     id = args[1]
     dn = None
  
  # identifying an optional argument with default value
  elements = 'no elements'
  if len(args) > 2:
    elements = args[1:]
```
Can be converted to
```
"""My script"""
from DIRAC.Core.Base import Script
from DIRAC.Core.Utilities.DIRACScript import DIRACScript

@DIRACScript()
def main():
  # it is important to add a colon after the name of the argument in the description
  Script.registerArgument(' Status: element status', values=['done', 'fail'])
  Script.registerArgument(('DN:  user DN', 'ID: user ID'))
  Script.registerArgument(['Element: list of elements'], default='no elements', mandatory=False)

  Script.parseCommandLine()
  status, second, elements = Script.getPositionalArguments(group=True)

  # detection of a controversial argument
  if second.startswith('/'):
    dn = second
    id = None
  else:
     id = second
     dn = None
```
The help output should be the same.

BEGINRELEASENOTES
Fragment from #5035:
- Added the ability to register positional arguments and add it to help message.
- Docs

*Core
NEW: provide some docs and add registerArgument method to register arguments in Script
FIX: align Script with changes in LocalConfiguration

*Configuration
NEW: add registerCmdArg to register arguments and group argument to group returned arguments
NEW: add Test_LocalConfiguration

*docs
FIX: rename exapmle scripts, docs

ENDRELEASENOTES

fstagni commented:
Looks OK to me.

andresailer commented:
Don't you want to add the use of `Script.registerArgument` to the two example scripts (ping_info, great_script)?
Also maybe move the example conversion out of the release notes?

chrisburr commented on src/DIRAC/Core/Utilities/DIRACScript.py:
```diff
@@ -45,7 +48,7 @@ def __call__(self, func=None):
     function_name = None
     for entrypoint in metadata.entry_points()['console_scripts']:
-      if not entrypoint.name.startswith("dirac-"):
+      if not entrypoint.name.startswith("dirac-") or self.scriptName != entrypoint.name:
```
I don't think it's a problem but out of curiosity, what's the reason for this?

TaykYoku replied:
you are right, in the context of this PR it makes no sense. I'll take it off. thanks

chrisburr commented on src/DIRAC/ConfigurationSystem/Client/test/Test_LocalConfiguration.py:
```diff
@@ -0,0 +1,218 @@
+
+@pytest.fixture
+def localCFG():
+  with mock.patch('DIRAC.ConfigurationSystem.Client.LocalConfiguration.gLogger.notice', side_effect=mock_notice):
```
I think you can use the [`monkeypatch` fixture](https://docs.pytest.org/en/stable/monkeypatch.html) to avoid needing lots of nested `with` statements.

TaykYoku replied:
ok

chrisburr commented on src/DIRAC/ConfigurationSystem/Client/LocalConfiguration.py:
```diff
@@ -310,6 +392,38 @@ def __parseCommandLine(self):
+      # Check accepted values
+      for cArg in self.commandArgList[i:i + 1 + step]:
+        if values and cArg not in values:
+          gLogger.fatal('Error when parsing command line arguments: \
+                         "%s" does not match the allowed values for %s' % (cArg, argMarking))
```
I think this will insert a big chunk of whitespace in the message? If so:
```suggestion
          gLogger.fatal(
              'Error when parsing command line arguments: '
              '"%s" does not match the allowed values for %s' % (cArg, argMarking)
          )
```

TaykYoku commented:
> Don't you want to add the use of `Script.registerArgument` to the two example scripts (ping_info, great_script)?

I added documentation to the appropriate methods in `LocalConfiguration` and updated the `dirac-my-great-script` script with examples of registration of various arguments.
Do we really need two examples of scripts? are they any different or should I add to `dirac-ping-info` the same as to the `dirac-my-great-script` script?

fstagni commented:
Can you rebase? Then we should hopefully see the tests not failing...?

TaykYoku commented:
PilotWrapper failed with:
```
2021-04-16 13:21:46 UTC dirac-install [NOTICE]  Installing DIRAC OS ...
2021-04-16 13:21:46 UTC dirac-install [DEBUG] Using DIRACOS tarball URL from configuration key Installations/DIRAC/DIRACOS
2021-04-16 13:21:46 UTC dirac-install [NOTICE]  Retrieving https://diracos.web.cern.ch/diracos/releases/diracos-master.tar.gz
2021-04-16 13:21:46 UTC dirac-install [DEBUG] Retrieving remote file "https://diracos.web.cern.ch/diracos/releases/diracos-master.tar.gz"
Error: -16 13:26:51 UTC dirac-install [ERROR] Cannot download diracos-master.tar.gz: Timeout
```
I have the same problem when testing locally via docker

fstagni commented:
> PilotWrapper failed with:
> 
> ```
> 2021-04-16 13:21:46 UTC dirac-install [NOTICE]  Installing DIRAC OS ...
> 2021-04-16 13:21:46 UTC dirac-install [DEBUG] Using DIRACOS tarball URL from configuration key Installations/DIRAC/DIRACOS
> 2021-04-16 13:21:46 UTC dirac-install [NOTICE]  Retrieving https://diracos.web.cern.ch/diracos/releases/diracos-master.tar.gz
> 2021-04-16 13:21:46 UTC dirac-install [DEBUG] Retrieving remote file "https://diracos.web.cern.ch/diracos/releases/diracos-master.tar.gz"
> Error: -16 13:26:51 UTC dirac-install [ERROR] Cannot download diracos-master.tar.gz: Timeout
> ```
> 
> I have the same problem when testing locally via docker

That's happening frequently these days. Just restart...
